src/software/ai/hl/stp/play/ball_placement/ball_placement_play_fsm.cpp

```diff
@@ -21,13 +22,69 @@ void BallPlacementPlayFSM::kickOffWall(const Update &event)
-    AutoChipOrKick auto_chick = {AutoChipOrKickMode::AUTOKICK,
-                                 WALL_KICKOFF_VELOCITY_M_PER_S};
+    Rectangle field_boundary     = event.common.world_ptr->field().fieldBoundary();
+
+    pickoff_final_orientation = calculateWallPickOffLocation(ball_pos, field_boundary, ROBOT_MAX_RADIUS_METERS * 4).first;
+    pickoff_point = ball_pos - Vector::createFromAngle(pickoff_final_orientation).normalize(0.4);
+    align_wall_tactic->updateControlParams(pickoff_point, pickoff_final_orientation,
+                                           0.0, TbotsProto::DribblerMode::OFF,
+                                           TbotsProto::BallCollisionType::AVOID, {AutoChipOrKickMode::OFF, 0},
+                                           TbotsProto::MaxAllowedSpeedMode::PHYSICAL_LIMIT,
+                                           TbotsProto::ObstacleAvoidanceMode::SAFE, 0.0);
+    tactics_to_run[0].emplace_back(align_wall_tactic);
+
+    event.common.set_tactics(tactics_to_run);
+}
+
+void BallPlacementPlayFSM::setPickOffDest(const BallPlacementPlayFSM::Update &event) {
+    Point ball_pos            = event.common.world_ptr->ball().position();
+    Rectangle field_boundary     = event.common.world_ptr->field().fieldBoundary();
+
+    std::pair<Angle, Point> location = calculateWallPickOffLocation(ball_pos, field_boundary, ROBOT_MAX_RADIUS_METERS * 4);
+
+    pickoff_final_orientation = location.first;
+    pickoff_destination = location.second - Vector::createFromAngle(pickoff_final_orientation).normalize(ROBOT_MAX_RADIUS_METERS * 5.5);
+}
+
+void BallPlacementPlayFSM::pickOffWall(const BallPlacementPlayFSM::Update &event) {
+    PriorityTacticVector tactics_to_run = {{}};
+
+    // setup move tactics for robots away from ball placing robot
+    setupMoveTactics(event);
+    tactics_to_run[0].insert(tactics_to_run[0].end(), move_tactics.begin(),
+                             move_tactics.end());
+
+    pickoff_wall_tactic->updateControlParams(pickoff_destination, pickoff_final_orientation, true, TbotsProto::MaxAllowedSpeedMode::BALL_PLACEMENT_WALL_DRIBBLE);
+
+    tactics_to_run[0].emplace_back(pickoff_wall_tactic);
+
+    event.common.set_tactics(tactics_to_run);
+}
+
+void BallPlacementPlayFSM::releaseBall(const BallPlacementPlayFSM::Update &event) {
+    PriorityTacticVector tactics_to_run = {{}};
+
+    // setup move tactics for robots away from ball placing robot
+    setupMoveTactics(event);
+    tactics_to_run[0].insert(tactics_to_run[0].end(), move_tactics.begin(),
+                             move_tactics.end());
+
+    Point ball_pos = event.common.world_ptr->ball().position();
+    std::optional<Robot> robot_placing_ball = event.common.world_ptr->friendlyTeam().getNearestRobot(ball_pos);
+
+    if (!robot_placing_ball.has_value())
+    {
+        return;
+    }
+
+    Point release_point = ball_pos + (robot_placing_ball->position() - ball_pos).normalize(.5);
```

Comment:
There's a lot of arbitrary constants around this file. We should either make them a constant in `ball_placement_play_fsm.h`, make them a variable with a descriptive name where it's used, or add a descriptive comment explaining the intuition behind the value.